### PR TITLE
sap.ui.fl as ui5 dependency and rolled-up

### PR DESCRIPTION
**Pull request by aleksandr-suvorov**
1. sap.ui.fl is is removed from project.json dependencies and now resolved by ui5 cli.
2. sap.ui.fl is bundled by rollup instead of using ui5 loader.
3. there are 2 rollup configs - one for production build, based on adaptation project, another uses project from test/resources/project for local build and test.

### CODE REVIEW

File: package.json [hunk 1]
```diff
@@ -27,21 +29,26 @@
...
     "dependencies": {
-        "@openui5/sap.ui.fl": "^1.81.0",
-        "@sap/cf-tools": "^0.6.8",
-        "@ui5/fs": "^2.0.3",
-        "@ui5/logger": "^2.0.0",
-        "adm-zip": "^0.4.14",
+        "@buxlabs/amd-to-es6": "^0.16.0",
+        "@rollup/plugin-node-resolve": "^11.1.0",
+        "@sap/cf-tools": "^0.8.0",
+        "@ui5/cli": "^2.8.0",
```
Comment (petermuessig) [inline package.json:35]: do you really need the @ui5/cli as dependency or should this rather be - you should be careful what you really need as dev dependency (for the local build) and what you need as regular dependency for the runtime of your code. E.g. also the rollup parts are IMO just devDependencies, right?

File: rollup/overrides/sap/ui/fl/Change.js [hunk 2]
```diff
@@ -0,0 +1,62 @@
+/*!
```
Comment (petermuessig) [inline rollup/overrides/sap/ui/fl/Change.js:1]: In some file we have a copyright, in some files not - at the UI5 tooling we do not have one - but I assume it is the case because you copied the source for the overlay from the OpenUI5 codebase - so maybe it is better to keep the copyright for this one...

File: rollup/rollupConfigBuilder.js [hunk 3]
```diff
@@ -0,0 +1,39 @@
...
+const { nodeResolve } = require("@rollup/plugin-node-resolve");
+
+class ConfigBuilder {
+    static build(projectPath) {
+        return {
+            input: "./rollup/bundle-def.js",
+            plugins: [
+                ui5({
+                    assets: [
+                        "/resources/sap/ui/fl/**"
```
Comment (petermuessig) [inline rollup/rollupConfigBuilder.js:13]: here you maintain the assets, but in the rollup plugin it is maintained again

File: rollup/ui5-resolve.js [hunk 4]
```diff
@@ -0,0 +1,118 @@
...
+
+        name: "ui5-resolve",
+
+        buildStart: async (buildOptions) => {
+            const project = await normalizer.generateProjectTree({
+                cwd: options.projectPath
+            });
+            this.dependencies = resourceFactory.createCollectionsForTree(project, {}).dependencies;
+
+            const resources = await this.dependencies.byGlob("/resources/sap/ui/fl/**");
```
Comment (petermuessig) [inline rollup/ui5-resolve.js:31]: here the assets are hard coded again - or did you want to use this configuration from assets?

File: rollup/ui5-resolve.js [hunk 5]
```diff
@@ -0,0 +1,118 @@
...
+                if (fs.existsSync(file)) {
+                    fs.unlinkSync(file);
+                }
+                return pipe(resource.getStream(), fs.createWriteStream(file));
+            });
+            await Promise.all(writePromises);
+        },
+
+
+        // Right before writting result to dist
```
Comment (petermuessig) [inline rollup/ui5-resolve.js:47]: typo "writing" only with one t

File: rollup/ui5-resolve.js [hunk 6]
```diff
@@ -0,0 +1,118 @@
...
+        renderChunk: async (code) => {
+            return "global.window = {};" + code;
```
Comment (petermuessig) [inline rollup/ui5-resolve.js:49]: why is this needed?

File: rollup/ui5-resolve.js [hunk 7]
```diff
@@ -0,0 +1,118 @@
...
+                return;
+            }
+            code = code
+                .replace(/sap\.ui\.define/g, "define")
+                .replace(/\, \/\* bExport\= \*\/ true\)/g, ")")
+                .replace(/}, true\);$/g, "});");
+
+            if (code) {
+                try {
+                    const result = amdextract.parse(code);
```
Comment (petermuessig) [inline rollup/ui5-resolve.js:105]: I think the unused import check is not needed for your scenario

you can also get rid of the amdextract dependency IMO

the complete try catch can be removed!

File: rollup/ui5-resolve.js [hunk 8]
```diff
@@ -0,0 +1,118 @@
...
+
+
+        // Right before writting result to dist
+        renderChunk: async (code) => {
+            return "global.window = {};" + code;
+        },
+
+
+        buildEnd: async (error) => {
+            log.verbose("buildEnd", error);
```
Comment (petermuessig) [inline rollup/ui5-resolve.js:54]: not needed for the productive plugin, right?

File: rollup/ui5-resolve.js [hunk 9]
```diff
@@ -0,0 +1,118 @@
+const path = require("path");
+const fs = require("fs");
+const stream = require("stream");
+const { promisify } = require("util");
+const pipe = promisify(stream.pipeline);
+
+const convertAMDtoES6 = require("@buxlabs/amd-to-es6");
+const amdextract = require("amdextract");
+
+const log = require("@ui5/logger").getLogger("rollup-plugin-ui5-resolve");
```
Comment (petermuessig) [inline rollup/ui5-resolve.js:10]: maybe rename it to `rollup-plugin-ui5-resolve-internal` since I might publish the `rollup-plugin-ui5-resolve` as a separate project to not collide - then you can also switch to the official one...

File: rollup/ui5-resolve.js [hunk 10]
```diff
@@ -0,0 +1,118 @@
+const path = require("path");
+const fs = require("fs");
+const stream = require("stream");
+const { promisify } = require("util");
+const pipe = promisify(stream.pipeline);
+
+const convertAMDtoES6 = require("@buxlabs/amd-to-es6");
+const amdextract = require("amdextract");
```
Comment (petermuessig) [inline rollup/ui5-resolve.js:8]: maybe not needed - see my comment below at the usage

File: rollup/ui5-resolve.js [hunk 11]
```diff
@@ -0,0 +1,118 @@
...
+
+const log = require("@ui5/logger").getLogger("rollup-plugin-ui5-resolve");
+const normalizer = require("@ui5/project").normalizer;
+const resourceFactory = require("@ui5/fs").resourceFactory;
+
+
+module.exports = (options) => {
+
+    log.verbose(JSON.stringify(options, undefined, 2));
```
Comment (petermuessig) [inline rollup/ui5-resolve.js:17]: necessary?

File: rollup/rollupConfigBuilder.js [hunk 12]
```diff
@@ -0,0 +1,39 @@
...
+                    ],
+                    exclude: [
+                        "./rollup/bundle-def.js",
+                        "sap/ui/thirdparty/URI"
```
Comment (petermuessig) [inline rollup/rollupConfigBuilder.js:17]: why do you have the overlay here and exclude it?

**Review by petermuessig — CHANGES_REQUESTED** [review]
I have some questions, see my comments in the code.

File: rollup/ui5-resolve.js (hunk 6 above)
Comment (aleksandr-suvorov) [reply rollup/ui5-resolve.js:49]: src/sap.ui.core/src/sap/base/util/ObjectPath.js requires a window object. I see 2 possible solutions, either .replace(/window/g, "{}") or add "window = {};" on top. The first solution can lead to unpredicted replacements, so I went with solution 2. Maybe you have a better idea how to handle it?

File: rollup/rollupConfigBuilder.js (hunk 12 above)
Comment (aleksandr-suvorov) [reply rollup/rollupConfigBuilder.js:17]: https://github.com/SAP/openui5/blob/master/src/sap.ui.core/src/sap/ui/thirdparty/URI.js was particulary difficult to convert. Either it required modification of the module, or override. Maybe the "exclude" word is misleading. It should be something like "skipTransformation".

File: package.json (hunk 1 above)
Comment (aleksandr-suvorov) [reply package.json]: Right, moved rollup dependencies to dev.

File: rollup/rollupConfigBuilder.js (hunk 3 above)
Comment (aleksandr-suvorov) [reply rollup/rollupConfigBuilder.js:13]: Missed that part earlier, now it's used in plugin.

File: rollup/ui5-resolve.js (hunk 8 above)
Comment (aleksandr-suvorov) [reply rollup/ui5-resolve.js]: Right ;)

File: rollup/ui5-resolve.js (hunk 9 above)
Comment (aleksandr-suvorov) [reply rollup/ui5-resolve.js]: Renamed to something like "rollup-plugin-ui5-resolve-task-adaptation" to distinguish better.

File: rollup/ui5-resolve.js (hunk 11 above)
Comment (aleksandr-suvorov) [reply rollup/ui5-resolve.js]: Good point, removed.

File: rollup/ui5-resolve.js (hunk 7 above)
Comment (aleksandr-suvorov) [reply rollup/ui5-resolve.js]: Less code is better ;)

File: rollup/rollupConfigBuilder.js (hunk 3 above)
Comment (petermuessig) [reply rollup/rollupConfigBuilder.js:13]: 👍

File: rollup/rollupConfigBuilder.js (hunk 12 above)
Comment (petermuessig) [reply rollup/rollupConfigBuilder.js:17]: 👍

File: test/resources/project/ui5.yaml [hunk 13]
```diff
@@ -0,0 +1,10 @@
+---
+specVersion: '2.2'
+type: application
+metadata:
+  name: HTML5Module
+framework:
+  name: OpenUI5
+  version: "1.85.2"
```
Comment (petermuessig) [inline test/resources/project/ui5.yaml:8]: Is this task version dependent ? Maybe we need to align the version with the runtime version then.

**Review by petermuessig — APPROVED** [review]
OK, just a topic to discuss => regarding the version alignment

**Comment by petermuessig** [comment]
Maybe @silkearians or @MatthiasSchmalz  also want to have a look into the change - high level it looks good now
